README.md
<h1 align="center">Bull Bus</h1>

<p align="center">
  Event Bus for Node.JS using <a href="https://github.com/OptimalBits/bull">Bull</a> Queues
</p>

<p align="center">
    <a href="https://github.com/AlbertHernandez/bull-bus/actions/workflows/nodejs.yml?branch=main"><img src="https://github.com/AlbertHernandez/bull-bus/actions/workflows/nodejs.yml/badge.svg?branch=main" alt="nodejs"/></a>
    <a href="https://www.npmjs.com/package/bull-bus"><img src="https://img.shields.io/npm/v/bull-bus.svg" alt="bull-bus-npm"/></a>
</p>

## Table of Contents

* [Installation](#installation)
* [How it works](#how-it-works)
* [How to Use It](#how-to-use-it)
    * [Bull Bus](#bull-bus)
    * [Bull Event Bus](#bull-event-bus)
    * [Visualization](#visualization)
* [Playground](#playground)
* [Preparing environment to contribute](#preparing-environment)
* [Building](#building)
* [Testing](#testing)
* [Linting](#linting)
* [Contributing](#contributing)

## Installation

```bash
npm install bull-bus
```

## How it works

When we work with event buses we normally have 1 event that can be consumed by N subscribers. When we want to create a new subscriber we will need to provide 3 main things:

* `Topic Name`: will be used to know the subscriptions that should be executed when a new topic is published.
* `Subscriber Name`: we can have N subscribers to a topic. The pair (topicName, subscriberName) will identify a unique subscription. Check how this is useful to [visualize](#visualization) the queues.
* `Handler`: this is the function that will be executed when an event is published to a particular topic.

## How to Use It

Bull Bus library offers two main functionalities. The bull bus and the bull event bus.

### Bull Bus

This class is a Bus Implementation using Bull, works with primitives data and does not know anything about the domain.
It may be useful in case we want to build our own domain event logic.

```ts
import { BullBus, Job } from "bull-bus";

const accountCreatedTopicName = "account-created";
const userCreatedTopicName = "user-created";

const sendEmailSubscriberName = "send-email";
const sendSlackSubscriberName = "send-slack";
const sendPushNotificationSubscriberName = "send-push-notification";

const bullBus = new BullBus({
  redisUrl: "redis://127.0.0.1:6379",
  topicNameToSubscriberNames: {
    [accountCreatedTopicName]: [
      sendEmailSubscriberName,
      sendSlackSubscriberName,
    ],
    [userCreatedTopicName]: [sendPushNotificationSubscriberName],
  },
});

interface AccountCreated {
  accountId: string;
}

interface UserCreated {
  userId: string;
}

bullBus.addSubscribers([
  {
    topicName: accountCreatedTopicName,
    handleEvent: async (job: Job<AccountCreated>) => {
      console.log(
        "Handle event account created, send email",
        job.data.accountId
      );
    },
    subscriberName: sendEmailSubscriberName,
  },
  {
    topicName: accountCreatedTopicName,
    handleEvent: async (job: Job<AccountCreated>) => {
      console.log(
        "Handle event account created, send slack",
        job.data.accountId
      );
    },
    subscriberName: sendSlackSubscriberName,
  },
  {
    topicName: userCreatedTopicName,
    handleEvent: async (job: Job<UserCreated>) => {
      console.log(
        "Handle event user created, send push notification",
        job.data.userId
      );
    },
    subscriberName: sendPushNotificationSubscriberName,
  },
]);

const accountCreatedEvent: AccountCreated = {
  accountId: "2",
};
const userCreatedEvent: UserCreated = {
  userId: "1",
};

await bullBus.publish(accountCreatedTopicName, accountCreatedEvent);
await bullBus.publish(userCreatedTopicName, userCreatedEvent);
```

### Bull Event Bus

Bull Event Bus is very similar to the Bull Bus with the difference that gives us some default classes to create domain
events and subscriptions. Its useful when we are working with OOP.

```ts
import {
  DomainEvent,
  DomainEventSubscriber,
  BullEventBus,
} from "bull-bus";

class UserRegistered extends DomainEvent {
  static EVENT_NAME = "user-registered";

  constructor(userName: string) {
    super({
      eventName: UserRegistered.EVENT_NAME,
      attributes: {
        userName,
      },
    });
  }
}

class UserFormCompleted extends DomainEvent {
  static EVENT_NAME = "user-form-completed";

  constructor(value: string) {
    super({
      eventName: UserFormCompleted.EVENT_NAME,
      attributes: {
        value,
      },
    });
  }
}

class SendSlackOnUserOrFormCompleted
  implements DomainEventSubscriber<UserRegistered | UserFormCompleted>
{
  subscribedTo() {
    return [UserRegistered, UserFormCompleted];
  }

  subscriberName(): string {
    return "send-slack";
  }

  async on(event: UserRegistered | UserFormCompleted) {
    switch (event.eventName) {
      case UserRegistered.EVENT_NAME:
        console.log("Simulating send slack...", event.attributes.userName);
        break;
      case UserFormCompleted.EVENT_NAME:
        console.log("Simulating send slack...", event.attributes.value);
        break;
    }
  }
}

class SendEmailOnUserRegistered
  implements DomainEventSubscriber<UserRegistered>
{
  subscribedTo() {
    return [UserRegistered];
  }

  subscriberName(): string {
    return "send-email";
  }

  async on(event: UserRegistered) {
    console.log("Simulating send email...", event.attributes.userName);
  }
}

const eventBus = new BullEventBus({
  redisUrl: "redis://127.0.0.1:6379",
  topicNameToSubscriberNames: {
    [UserRegistered.EVENT_NAME]: ["send-slack", "send-email"],
    [UserFormCompleted.EVENT_NAME]: ["send-slack"],
  },
});

eventBus.addSubscribers([
  new SendSlackOnUserOrFormCompleted(),
  new SendEmailOnUserRegistered(),
]);

await eventBus.publish([new UserRegistered("gabriel")]);
await eventBus.publish([new UserFormCompleted("3208")]);
```

### Visualization

Both buses are ready to show the internal queues to display the job data in a pretty way. The following image is using <a href="https://taskforce.sh/">Taskforce</a>, but can be used any UI for Bull.

<a href="https://github.com/AlbertHernandez/insomnia-plugin-ocean-light-theme">
    <img src="images/taskforce.png" alt="Logo">
</a>

## Playground

This library offers a playground where we can play with the functions that we are developing

```bash
docker-compose up -d redis
npm run playground
```

## Preparing environment to contribute

This library has been designed to work with node v16 and npm 8. In order to configure your local environment you can
run:

```bash
nvm install 16.0.0
nvm use
npm install npm@8.3.0 -g
npm install
```

## Building

```bash
npm run build
```

## Testing

### Jest with Testing Library

```bash
npm run test
```

## Linting

Run the linter

```bash
npm run lint
```

Fix lint issues automatically

```bash
npm run lint:fix
```

## Contributing

Contributions welcome! See
the [Contributing Guide](https://github.com/AlbertHernandez/bull-bus/blob/main/CONTRIBUTING.md).


CHANGELOG.md
# Changelog

### [1.4.1](https://www.github.com/AlbertHernandez/bull-bus/compare/v1.4.0...v1.4.1) (2022-01-26)


### Bug Fixes

* improve wrong playground and readme example ([77dbfce](https://www.github.com/AlbertHernandez/bull-bus/commit/77dbfceba05be18ee024f82f578ffa626c6def11))

## [1.4.0](https://www.github.com/AlbertHernandez/bull-bus/compare/v1.3.1...v1.4.0) (2022-01-23)


### Features

* improve types and docs ([34e56ee](https://www.github.com/AlbertHernandez/bull-bus/commit/34e56eee886d41309b23f18762b4fd8d60afe0d2))


### Bug Fixes

* add imports in the readme ([9da69c6](https://www.github.com/AlbertHernandez/bull-bus/commit/9da69c6d6ca5be7ccdd26ea63b23adddcc7c8fb8))
* reorder readme ([9539b49](https://www.github.com/AlbertHernandez/bull-bus/commit/9539b49190fae2f75006d00cf60524238421c257))

### [1.3.1](https://www.github.com/AlbertHernandez/bull-bus/compare/v1.3.0...v1.3.1) (2022-01-23)


### Bug Fixes

* improve readme title ([0e96fde](https://www.github.com/AlbertHernandez/bull-bus/commit/0e96fdee20d98ec44c6f30d414ca7029ac10e92d))

## [1.3.0](https://www.github.com/AlbertHernandez/bull-bus/compare/v1.2.0...v1.3.0) (2022-01-23)


### Features

* allow to send event options ([572c07e](https://www.github.com/AlbertHernandez/bull-bus/commit/572c07e87aed7c02f94537eace52093030b3db54))

## [1.2.0](https://www.github.com/AlbertHernandez/bull-bus/compare/v1.1.1...v1.2.0) (2022-01-23)


### Features

* allow publish events when there is no subscribers ready ([44a50f4](https://www.github.com/AlbertHernandez/bull-bus/commit/44a50f4d23c7fb1261ffea4a776c01e4211d266b))

### [1.1.1](https://www.github.com/AlbertHernandez/bull-bus/compare/v1.1.0...v1.1.1) (2022-01-22)


### Bug Fixes

* improve readme and links ([1834166](https://www.github.com/AlbertHernandez/bull-bus/commit/18341664d088b66bc8cb78d46d22cbbfb3b5b817))

## [1.1.0](https://www.github.com/AlbertHernandez/bull-bus/compare/v1.0.1...v1.1.0) (2022-01-22)


### Features

* use subscriber name to get queue name ([9f426cd](https://www.github.com/AlbertHernandez/bull-bus/commit/9f426cd284ac4e173fa400a08e2f1ac16c8cdf23))

### [1.0.1](https://www.github.com/AlbertHernandez/bull-bus/compare/v1.0.0...v1.0.1) (2022-01-22)


### Bug Fixes

* improve docs and license ([b5dc332](https://www.github.com/AlbertHernandez/bull-bus/commit/b5dc3327ec7f20950408e92b6ad587c09a7059e5))

## 1.0.0 (2022-01-22)


### Features

* create initial version ([40e78f9](https://www.github.com/AlbertHernandez/bull-bus/commit/40e78f933c5fe409c01509eff9a4c0a983c4fad9))


playground/index.ts
import {
  DomainEvent,
  DomainEventSubscriber,
  BullEventBus,
  BullBus,
  Job,
} from "../src";

const runBullEventBus = async () => {
  class UserRegistered extends DomainEvent {
    static EVENT_NAME = "user-registered";

    constructor(userName: string) {
      super({
        eventName: UserRegistered.EVENT_NAME,
        attributes: {
          userName,
        },
      });
    }
  }

  class UserFormCompleted extends DomainEvent {
    static EVENT_NAME = "user-form-completed";

    constructor(value: string) {
      super({
        eventName: UserFormCompleted.EVENT_NAME,
        attributes: {
          value,
        },
      });
    }
  }

  class SendSlackOnUserOrFormCompleted
    implements DomainEventSubscriber<UserRegistered | UserFormCompleted>
  {
    subscribedTo() {
      return [UserRegistered, UserFormCompleted];
    }

    subscriberName(): string {
      return "send-slack";
    }

    async on(event: UserRegistered | UserFormCompleted) {
      switch (event.eventName) {
        case UserRegistered.EVENT_NAME:
          console.log("Simulating send slack...", event.attributes.userName);
          break;
        case UserFormCompleted.EVENT_NAME:
          console.log("Simulating send slack...", event.attributes.value);
          break;
      }
    }
  }

  class SendEmailOnUserRegistered
    implements DomainEventSubscriber<UserRegistered>
  {
    subscribedTo() {
      return [UserRegistered];
    }

    subscriberName(): string {
      return "send-email";
    }

    async on(event: UserRegistered) {
      console.log("Simulating send email...", event.attributes.userName);
    }
  }

  const eventBus = new BullEventBus({
    redisUrl: "redis://127.0.0.1:6379",
    topicNameToSubscriberNames: {
      [UserRegistered.EVENT_NAME]: ["send-slack", "send-email"],
      [UserFormCompleted.EVENT_NAME]: ["send-slack"],
    },
  });

  eventBus.addSubscribers([
    new SendSlackOnUserOrFormCompleted(),
    new SendEmailOnUserRegistered(),
  ]);

  await eventBus.publish([new UserRegistered("gabriel")]);
  await eventBus.publish([new UserFormCompleted("3208")]);
};

const runBullBus = async () => {
  const accountCreatedTopicName = "account-created";
  const userCreatedTopicName = "user-created";

  const sendEmailSubscriberName = "send-email";
  const sendSlackSubscriberName = "send-slack";
  const sendPushNotificationSubscriberName = "send-push-notification";

  const bullBus = new BullBus({
    redisUrl: "redis://127.0.0.1:6379",
    topicNameToSubscriberNames: {
      [accountCreatedTopicName]: [
        sendEmailSubscriberName,
        sendSlackSubscriberName,
      ],
      [userCreatedTopicName]: [sendPushNotificationSubscriberName],
    },
  });

  interface AccountCreated {
    accountId: string;
  }

  interface UserCreated {
    userId: string;
  }

  bullBus.addSubscribers([
    {
      topicName: accountCreatedTopicName,
      handleEvent: async (job: Job<AccountCreated>) => {
        console.log(
          "Handle event account created, send email",
          job.data.accountId
        );
      },
      subscriberName: sendEmailSubscriberName,
    },
    {
      topicName: accountCreatedTopicName,
      handleEvent: async (job: Job<AccountCreated>) => {
        console.log(
          "Handle event account created, send slack",
          job.data.accountId
        );
      },
      subscriberName: sendSlackSubscriberName,
    },
    {
      topicName: userCreatedTopicName,
      handleEvent: async (job: Job<UserCreated>) => {
        console.log(
          "Handle event user created, send push notification",
          job.data.userId
        );
      },
      subscriberName: sendPushNotificationSubscriberName,
    },
  ]);

  const accountCreatedEvent: AccountCreated = {
    accountId: "2",
  };
  const userCreatedEvent: UserCreated = {
    userId: "1",
  };

  await bullBus.publish(accountCreatedTopicName, accountCreatedEvent);
  await bullBus.publish(userCreatedTopicName, userCreatedEvent);
};

runBullBus();
runBullEventBus();
